src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorAdornmentManager.cs

```diff
@@ -0,0 +1,216 @@
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineErrors
+{
+    internal class InlineErrorAdornmentManager : AdornmentManager<InlineErrorTag>
+    {
+        private readonly IClassificationTypeRegistryService _classificationRegistryService;
+        private readonly IClassificationFormatMap _formatMap;
+        private readonly Dictionary<IMappingTagSpan<InlineErrorTag>, SnapshotPoint> _tagSpanToPointMap;
+
+        public InlineErrorAdornmentManager(IThreadingContext threadingContext,
+            IWpfTextView textView, IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IAsynchronousOperationListener asyncListener, string adornmentLayerName,
+            IClassificationFormatMapService classificationFormatMapService,
+            IClassificationTypeRegistryService classificationTypeRegistryService)
+            : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
+        {
+            _classificationRegistryService = classificationTypeRegistryService;
+            _formatMap = classificationFormatMapService.GetClassificationFormatMap(textView);
+            _formatMap.ClassificationFormatMappingChanged += OnClassificationFormatMappingChanged;
+            _tagSpanToPointMap = new Dictionary<IMappingTagSpan<InlineErrorTag>, SnapshotPoint>();
+        }
+
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            if (AdornmentLayer is not null)
+            {
+                var elements = AdornmentLayer.Elements;
+                foreach (var element in elements)
+                {
+                    var tag = (InlineErrorTag)element.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineErrorTag.TagID + tag.ErrorType);
+                    var format = GetFormat(classificationType);
+                    tag.UpdateColor(format, element.Adornment);
+                }
+            }
+        }
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that I can only display the first one that appears on the line
+        /// </summary>
+        private ImmutableDictionary<int, List<IMappingTagSpan<InlineErrorTag>>> GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            _tagSpanToPointMap.Clear();
+            if (changedSpanCollection.IsEmpty())
+            {
+                return ImmutableDictionary<int, List<IMappingTagSpan<InlineErrorTag>>>.Empty;
```

Comment:
fwiw, an immutable dictionary that points to mutable lists is a bit weird.  i'd just go with a Dictionary here if this is all just temp state that is passed out.

Reply:
changed